### PR TITLE
(PDB-1197) anonymize logs and metrics

### PR DESCRIPTION
**Pull request by wkalt**
This implements anonymization for logs and metrics.

### CODE REVIEW

File: src/puppetlabs/puppetdb/anonymizer.clj [hunk 1]
```diff
@@ -350,6 +359,38 @@
         (update-in-nil ["containment_path"]
                        anonymize-containment-path newcontext config))))
 
+(pls/defn-validated anonymize-metric :- metric-schema-str
+  [metric :- metric-schema-str
+   context
+   config]
+  (-> metric
+      (update-in ["name"] anonymize-leaf :name context config)
+      (update-in ["category"] anonymize-leaf :name context config)
```
Comment (kbarber) [inline src/puppetlabs/puppetdb/anonymizer.clj]: Does category even need anonymization?

File: src/puppetlabs/puppetdb/anonymizer.clj [hunk 2]
```diff
@@ -350,6 +359,38 @@
         (update-in-nil ["containment_path"]
                        anonymize-containment-path newcontext config))))
 
+(pls/defn-validated anonymize-metric :- metric-schema-str
+  [metric :- metric-schema-str
+   context
+   config]
+  (-> metric
+      (update-in ["name"] anonymize-leaf :name context config)
```
Comment (kbarber) [inline src/puppetlabs/puppetdb/anonymizer.clj]: Sometimes the 'name' presents a resource type ... this should follow the same anonymization rules as resource type elsewhere to keep the shape the same.

**Comment by kbarber** [comment]
I notice there is no policy additions to handle for example, logs ... seems like we've added a couple of things that should be switchable. For example the metric 'values' aren't always important to anonymized.

**Comment by wkalt** [comment]
@pljenkinsro retest this please

File: src/puppetlabs/puppetdb/anonymizer.clj [hunk 3]
```diff
@@ -350,6 +366,37 @@
         (update-in-nil ["containment_path"]
                        anonymize-containment-path newcontext config))))
 
+(pls/defn-validated anonymize-metric
+  [metric :- metric-schema-str
+   context
+   config]
+  (if (= "time" (get metric "category"))
+    (update-in metric ["name"] anonymize-leaf :type context config)
```
Comment (kbarber) [inline src/puppetlabs/puppetdb/anonymizer.clj]: So the metric resource name under `time` is actually in lower case, while the :type stuff expects the first letter to be upper case. This is creating a difference between the events and metrics area, ie. the random strings don't match - not even if you make the first letter lower case later on. I think you need to upcase the first letter of the input before getting something back, then lower it again or something like that - haven't really thought it through, perhaps something a bit more elegant :-).

File: documentation/anonymization.markdown [hunk 4]
```diff
@@ -45,7 +45,7 @@ At the moment there are some different built-in profiles you can choose from, an
 
 ### Profile: full
 
-The `full` profile will anonymize everything, while keeping the shape of data as previously mentioned. This includes: node names, resource types, resource titles, parameter names, values, any log messages, file names and file lines. The result should be a completely anonymized data set.
+The `full` profile will anonymize everything, while keeping the shape of data as previously mentioned. This includes: node names, resource types, resource titles, parameter names, values, any log messages, file names and file lines. The result should be a completely anonymized data set. Report metrics under the resources and events categories are left intact, as these can be inferred from the rest of the data, but names of metrics under the 'time' category are anonymized.
```
Comment (kbarber) [inline documentation/anonymization.markdown]: Backtick 'time'. And maybe resources and events also?

File: src/puppetlabs/puppetdb/anonymizer.clj (hunk 3 above)
Comment (kbarber) [inline src/puppetlabs/puppetdb/anonymizer.clj]: I think tags had a similar issue, take a look at what its doing.

File: documentation/anonymization.markdown [hunk 5]
```diff
@@ -63,6 +63,8 @@ The `moderate` profile attempts to be a bit smarter about what it anonymizes and
 * report log messages: are always anonymized
 * file names: are always anonymized
 * file numbers: are left as they are
+* metrics: are left intact
```
Comment (kbarber) [inline documentation/anonymization.markdown]: Well this isn't true, the `time` category gets anonymized based on the resource type policy above.

**Comment by wkalt** [comment]
@kbarber thanks, fixed

**Comment by wkalt** [comment]
@pljenkinsro retest this please

**Comment by wkalt** [comment]
@pljenkinsro retest this please

**Pull request merged**